docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,263 @@
+
+C++ provides tools to create high-performance APIs. C++ interoperability should embrace this. Interop should not be a significant source of overhead and performance concerns should not be a reason to continue using C++ to call C++ APIs rather than Swift.
+
+C++ is an un-opinionated, multi-paradigm language, designed to fit many use-cases. Different codebases often express the same concept in different ways. There is no prevailing consensus among C++ programmers about the right way to express specific concepts: how to name types and methods, how much to use templates, when to use heap allocation, how to propagate and handle errors, and so on. This creates problems for importing C++ APIs into Swift, which tends to have stronger conventions, some of which are backed by language rules. For instance, it is a common pattern in some C++ codebases to have classes that are only (or at least mostly) intended to be heap-allocated and passed around by pointer; consider this example:
+
+```
+// StatefulObject has object identity and reference semantcs: 
+// it should be constructed with "create" and used via a pointer.
+struct StatefulObject {
+  StatefulObject(const StatefulObject&) = delete;
+  StatefulObject() = delete;
+
+  StatefulObject *create() { return new StatefulObject(); }
+};
+```
+
+This type is not intended to be used directly as the type of a local variable or a `std::vector` element; it is allocated on the heap by the `create` method and passed around as a pointer. This is weakly enforced by the way it hides its constructors, but mostly it's communicated in the documentation and by the overall shape of the API; there is no language feature or programming pattern that expresses the specific semantics of this type. 
+
+In Swift, `StatefulObject` would be explicitly defined as a reference type using a specific language construct (defining it as a `class`). Swift defines clear patterns for naming, generic programming, value categories, error handling, and so on, and codebases are encouraged to use these standard practices. These well-defined programming patterns make using Swift APIs a cohesive experience; an experience that C++ interoperability should stive to maintain for Swift programmers using C++ APIs.
```

Comment:
```suggestion
If `StatefulObject` were written in Swift, it would be idiomatically be defined as a `class` to make it a reference type. This is an example of how Swift defines clear patterns for naming, generic programming, value categories, error handling, and so on, which codebases are encouraged to use as standard practices. These well-defined programming patterns make using Swift APIs a cohesive experience, and C++ interoperability should strive to maintain that for Swift programmers using C++ APIs.
```